networkit/cpp/embedding/BiasedRandomWalk.cpp

```diff
@@ -0,0 +1,200 @@
+ *  part of snap [https://github.com/snap-stanford/snap]
+ *  Copyright (c) 2007-2019, Jure Leskovec (under BSD license)
+ *
+ *  see [https://arxiv.org/pdf/1607.00653v1.pdf]
+ *
+ */
+
+// networkit-format
+
+#include <algorithm>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+#include <networkit/auxiliary/Random.hpp>
+#include <networkit/graph/Graph.hpp>
+
+#include "AliasSampler.hpp"
+#include "BiasedRandomWalk.hpp"
+
+namespace NetworKit {
+namespace Embedding {
+
+using NeighborSet = std::unordered_set<node>;
+using NeighborMap = std::unordered_map<node, AliasSampler>;
+
+struct GraphData {
+    // assuming contiguous node numbers 0..n-1, they serve as indices
+    using Data = std::vector<NeighborMap>;
+    Data data;
+
+    GraphData(size_t nn) : data{nn, NeighborMap()} {}
+};
+
+std::unique_ptr<GraphData> graphData;
+
+void preprocessNode(const Graph &graph, node t, double paramP, double paramQ) {
+
+    // for node t
+    NeighborSet tNbrs; // Neighbors of t
+    for (auto i : Graph::NeighborRange<>(graph, t)) {
+        tNbrs.insert(i);
+    }
+    // for each t-neighbor v:
+    for (auto v : Graph::NeighborRange<>(graph, t)) {
+        double pSum = 0;
+        std::vector<float> pTable; // Probability distribution table
+        // for each v-neighbor 
+        graph.forNeighborsOf(v, [&](node x, edgeweight  weight) {
+            if (x == t) {
+                pTable.push_back(weight / paramP);
+                pSum += weight / paramP;
+            } else if (tNbrs.count(x)) {
+                pTable.push_back(weight);
+                pSum += weight;
+            } else {
+                pTable.push_back(weight / paramQ);
+                pSum += weight / paramQ;
+            }
+        });
+        // Normalizing table
+        float pfSum = (float)pSum;
+        std::for_each(pTable.begin(), pTable.end(), [pfSum](float &p) { p /= pfSum; });
+
+        graphData->data[v][t].unigram(pTable);
+    }
+}
+
+std::vector<std::vector<node>> index2node;
+
+// Preprocess transition probabilities for each path t->v->x
+void preprocessTransitionProbs(const Graph &graph, double paramP, double paramQ) {
+    auto nn = graph.numberOfNodes();
+    // graphData = std::make_unique<GraphData>(nn);
+    // back to c++11 ;-(
+    graphData = std::unique_ptr<GraphData>(new GraphData(nn));
+
+    index2node.resize(nn);
+
+    // pre-allocate unordered maps:
+    for (index v = 0; v < nn; ++v) {
+        auto degv = graph.degreeOut(v);
+        auto neighbors = Graph::NeighborRange<>(graph, v);
+        for (auto n : neighbors) {
+            // init index2node:
+            index2node[v].push_back(n);
+            graphData->data[v][n] = AliasSampler(degv);
+        }
+    }
+
+    count nCnt = 0;
+
+#pragma omp parallel for schedule(dynamic)
+    // is forNodes parallelized for omp ?
+    //  for (node t = 0; t < nn; ++t) {
+    for (omp_index t = 0; t < static_cast<omp_index>(nn); ++t) {
```

Comment:
The node ids of a graph are not guaranteed to be continue, use `G->balancedParallelForNodes([&](node u) { /* ... */ });` ?